### ISSUE TITLE
Segmentation fault after processing data in a different process

### ISSUE THREAD

**Issue by joachimwolff** [pysam-developers/pysam#425]
I have the following issue and do not know if it is my fault or an error in pysam:

I read the data from a samfile, store it in a buffer and then I hand over this buffer to a new Process of the multiprocessing library in python. Inside this new process I can access the data without any issue but if I return via a Queue a new buffer which contains elements of the old buffer if some requirements are fullfilled, I get a segmentation fault. 

I will show you a code example and I hope you can understand better what is going on.

```python

import pysam
import multiprocessing

def readData(pFileOneIterator, pFileTwoIterator, pNumberOfItemsPerBuffer):
   j = 0
   buffer_1 = []
   buffer_2 = []
   while j < pNumberOfItemsPerBuffer:
        j += 1
        try:
            mate1 = pFileOneIterator.next()
            mate2 = pFileTwoIterator.next()
        except StopIteration:
            break
        # if some_conditions on mate1 and mate2 fit
        buffer_1.append(mate1)
        buffer_2.append(mate2)
    return buffer_1, buffer_2

def process_data(pMateBuffer1, pMateBuffer2, pQueue):
    # process data
    i = 0
    buffer_out = []
    while i < len(pMateBuffer1):
        mate1 = pMateBuffer1[i]
        mate2 = pMateBuffer2[i]

        # if some conditions are fulfilled, add 
        buffer_out.append(mate1)
        buffer_out.append(mate2)
   # this access works
   print buffer_out[0].flag
   pQueue.put(buffer_out)
  
def main(args=None):
    # create alignment file
    str1 = pysam.Samfile('File1', 'rb')
    str2 = pysam.Samfile('File2', 'rb')

    buffer_1, buffer_2 = readData(str1, str2, 500000)
    queue = multiprocessing.Queue()
    p = multiprocessing.Process(target=process_data, kwargs=dict(
                    pMateBuffer1=buffer_1,
                    pMateBuffer2=buffer_2,
                    pQueue=queue
                    ))
    p.start()
    # let process compute and then get data back
    result = queue.get()
    # type can be seen:
    print type(result[0]) # prints <type 'pysam.libcalignedsegment.AlignedSegment'>
    # but this access raises a segementation fault
    print result[0].flag # Segmentation fault (core dumped)
    p.join()
```
Why can I copy the data from process A to B, but for some reasons it fails to copy all the data back from B to A? Am I doing something wrong if yes, what? For me it looks like that some pointers or references in the underlying data structure are deleted. 

Thanks a lot for your help!

**Comment by kghosesbg** [comment]
I've always had strange issues (like segmentation faults) when most pysam objects cross process boundaries. This is especially true of AlignedSegment() objects (I believe these carry references to the original files). 

My suggestion would be to open the pysam file, read it and operate on the objects all in the same process and not pass pysam objects between processes.

**Comment by joachimwolff** [comment]
@kghosesbg Thanks for your answer! 

Unfortunately to not pass pysam objects between processes is not a solution for me. I solved it in the meantime by writing the output to /dev/shm (a ram disk which is supported by all (?) modern Linux kernels) and running an additional process in the background which is merging these files.  After the computation is done it writes it to the desired location and deletes the file at /dev/shm.

**Comment by kghosesbg** [comment]
@joachimwolff ah, very neat. So you communicate (as far as the code is concerned) by reading and writing reads to a BAM file?

**Comment by joachimwolff** [comment]
@kghosesbg  My plan was first to write the pysam objects in each subprocess to a buffer and if the buffer is full to write all elements of this buffer out to a BAM file, but this step should be done by the main process to prevent multiple access to the output which is a slowdown. As mentioned above this is not working. To overcome this limitation I write now in each process a BAM file to /dev/shm (aka my 'buffer') and run an additional process which is merging the BAM files (close the BAM file if the 'buffer' is full) into one big BAM file. 
I do not really communicate with the BAM files between the processes, I just write the result of each process out to a bam file. After this is done I have multiple BAM files and I merge them. If you are interested in the real implementation have a look: https://github.com/joachimwolff/HiCExplorer/blob/profile_performance/hicexplorer/hicBuildMatrix.py. 
Of interest is: Line 914, 915, function 'write_bam', line 925 which is running with a background process.

**Comment by kghosesbg** [comment]
@joachimwolff I've used a slightly different approach. I can't speak as to the efficiency re: your method, but here it is: I use the `pysam.AlignedSegment.tostring()` method to convert the read into a string, which can then safely pass the process boundary. On the other side I use a fromstring method (that I wrote - pysam does not supply one, interestingly) to retrieve the read object.

In my use case, however, the only time reads have to cross process boundaries is when they are part of a read pair where a mate is very far away, so the number of such crossings is relatively low.
